server/src/main/scala/services/prediction/SplitsPredictor.scala

```diff
@@ -0,0 +1,206 @@
+
+  lazy val flightCodesToTrain: Seq[String] = flightsHavingTrainingExamples
+
+  lazy val features: IndexedSeq[String] = extractFeatures
+
+  lazy val paxTypeModels: Map[String, LinearRegressionModel] = paxTypes.map(paxType => (paxType, trainModel(paxType, features))).toMap
+
+  def predictForArrivals(arrivals: Seq[Arrival]): Seq[(Arrival, Option[ApiSplits])] = {
+    val arrivalsToPredict = arrivals.filter(a => flightCodesToTrain.contains(a.IATA))
+
+    val flightSplitsPredictions: List[((String, Long), String, Double)] = paxTypes.flatMap(paxType => {
+      log.info(s"Predicting $paxType values for ${arrivalsToPredict.length} arrivals")
+
+      val predictionDf: DataFrame = predictionSetFromArrivals(arrivalsToPredict, features)
+
+      val withPrediction = paxTypeModels(paxType).transform(predictionDf)
+
+      val predictions: Seq[((String, Long), String, Double)] = withPrediction
+        .collect.toSeq
+        .map(row => {
+          val flightCode = row.getAs[String]("flight")
+          val scheduled = row.getAs[Long]("scheduled")
+          val prediction = row.getAs[Double]("prediction")
+
+          ((flightCode, scheduled), paxType, prediction)
+        })
+
+      log.info(s"Predicted ${predictions.length} $paxType values")
+      predictions
+    })
+
+    log.info(s"Predicted ${flightSplitsPredictions.length} splits for ${arrivalsToPredict.length} arrivals")
+
+    arrivalsWithPredictions(arrivals, flightSplitsPredictions)
+  }
+
+  def arrivalsWithPredictions(arrivals: Seq[Arrival], flightSplitsPredictions: Seq[((String, Long), String, Double)]): Seq[(Arrival, Option[ApiSplits])] = arrivals
+    .map(a => {
+      val arrivalSplits: Seq[(String, Double)] = flightSplitsPredictions
+        .collect {
+          case ((flightCode, scheduled), splitType, prediction) if flightCode == a.IATA && scheduled == a.Scheduled => (splitType, prediction)
+        }
+      val predictedSplits = arrivalSplits match {
+        case splits if splits.isEmpty => None
+        case splits =>
+          val paxTypeAndQueueCounts = splits
+            .map {
+              case (splitType, prediction) =>
+                ApiPaxTypeAndQueueCount(PaxType(s"""$splitType$$"""), splitTypeToQueue(splitType), prediction, None)
+            }
+            .toSet
+          Option(ApiSplits(paxTypeAndQueueCounts, SplitSources.PredictedSplitsWithHistoricalEGateAndFTPercentages, None, Percentage))
+      }
+      (a, predictedSplits)
+    })
+
+  def predictionSetFromArrivals(arrivalsToPredict: Seq[Arrival], features: IndexedSeq[String]): DataFrame = {
+    import sparkSession.implicits._
+
+    val validationSetDf = arrivalsToPredict
+      .map(arrival => {
+        val sf: Seq[(Int, Double)] = featureSpecs
+          .map { fs =>
+            val featureValue = fs.columns.map(c => arrivalFeature(c, arrival)).mkString("-")
+            val featureString = s"${fs.featurePrefix}$featureValue"
+            val featureIdx = features.indexOf(featureString)
```

Comment:
check if there's something that returns an option here.

Reply:
there isn't :(

Reply:
Could zipWithIndex and then use find() but that wouldn't necessarily improve clarity